File: parent/core/src/main/java/net/sf/jguiraffe/gui/builder/components/tags/ListModelSupport.java
/*
 * Copyright 2006-2015 The JGUIraffe Team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.sf.jguiraffe.gui.builder.components.tags;

import net.sf.jguiraffe.gui.builder.components.model.ListModel;

import org.apache.commons.jelly.JellyContext;

/**
 * <p>
 * Definition of an interface for tags that support a list model.
 * </p>
 * <p>
 * This interface will be implemented by tag handler classes that create
 * components like list boxes or combo boxes that are associated with a
 * {@link ListModel}. Other tags that create such models will look for enclosing
 * tags implementing this interface to set the newly created models.
 * </p>
 * <p>
 * This interface is also used for resolving the list model and looking it up in
 * a {@code BeanContext}. Therefore there are some other methods for querying
 * information related to the list model.
 * </p>
 *
 * @author Oliver Heger
 * @version $Id: ListModelSupport.java 205 2012-01-29 18:29:57Z oheger $
 */
public interface ListModelSupport
{
    /**
     * Returns the current {@code ListModel}. This method can return <b>null</b>
     * if no model has been set so far.
     *
     * @return the current {@code ListModel}
     */
    ListModel getListModel();

    /**
     * Sets the {@code ListModel}.
     *
     * @param model the {@code ListModel} to be set
     */
    void setListModel(ListModel model);

    /**
     * Returns the reference name of the list model. This name is used for
     * looking up the {@code ListModel} in the current {@code BeanContext}.
     *
     * @return the name of the bean representing the list model
     */
    String getModelRef();

    /**
     * Returns the current Jelly context. This object is needed for performing a
     * lookup to find the list model object.
     *
     * @return the current Jelly context
     */
    JellyContext getContext();
}

File: parent/java-fx/src/main/scala/net/sf/jguiraffe/gui/platform/javafx/builder/components/ListModelSupport.scala
/*
 * Copyright 2006-2017 The JGUIraffe Team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.sf.jguiraffe.gui.platform.javafx.builder.components

import javafx.collections.ObservableList
import net.sf.jguiraffe.gui.builder.components.model.ListComponentHandler
import net.sf.jguiraffe.gui.builder.components.model.ListModel
import net.sf.jguiraffe.gui.builder.components.tags.ListModelUtils

/**
 * Definition of a trait supporting component handlers which manage controls
 * with a list model.
 *
 * This trait can be mixed in handlers for combo boxes and lists. It implements
 * functionality for managing a list model, including adding or removing model
 * elements.
 *
 * This implementation expects that the ''initListModel()'' is called before
 * the object is actually used. Because it is used internally only, no
 * sophisticated checks are in place.
 */
private trait ListModelSupport extends ListComponentHandler {
  /** The ''EditableComboBoxModel'' provided by the associated model. */
  lazy val editableModel = initEditableComboBoxModel()

  /**
   * The list to be used for backing the ''JavaFxListModel'' object's display
   * list. This list is associated with a JavaFX control.
   */
  protected val displayList: ObservableList[Object]

  /** Stores the original data model from which this object is initialized. */
  private var originalListModel: ListModel = _

  /** Stores the FX list model used by this object. */
  private var fxListModel: JavaFxListModel = _

  /**
   * Initializes the ''JavaFxListModel'' managed by this object from the passed
   * in generic list model.
   * @param dataModel the underlying list model
   */
  def initListModel(dataModel: ListModel) {
    originalListModel = dataModel
    fxListModel = new JavaFxListModel(displayList, dataModel.getType)
    fxListModel.initFromModel(dataModel)
  }

  def getListModel: ListModel = fxListModel

  /**
   * @inheritdoc This implementation adds the new item to the FX list model.
   */
  def addItem(index: Int, display: Object, value: Object) {
    fxListModel.insertItem(index, display, value)
  }

  /**
   * @inheritdoc This implementation removes the specified element from the FX
   * list model.
   */
  def removeItem(index: Int) {
    fxListModel.removeItem(index)
  }

  /**
   * Initializes an ''EditableComboBoxModel'' from the original list model.
   * @return the ''EditableComboBoxModel''
   */
  private def initEditableComboBoxModel() =
    ListModelUtils.fetchEditableComboBoxModel(originalListModel)
}
